### PR TITLE
Implement saturating_abs() and saturating_neg() functions for signed integer types

### LINKED ISSUE #59983: Add saturating_abs and saturating_neg
It would be great to have `saturating_abs` and `saturating_neg` functions for signed integer types. The current work-around is using a pattern like `checked_$opr().unwrap_or($typ::max_value())` but having explicit functions would lift the corner-case knowledge "burden" from the user.

### PR DESCRIPTION
**Pull request by t-rapp**
Similar to wrapping_abs() / wrapping_neg() functions but saturating at the numeric bounds instead of wrapping around. Complements the existing set of functions with saturation mechanics.

cc #59983

### CODE REVIEW

**Comment by hellow554** [comment]
Please add some tests (especially the corner cases of course) in `src/libcore/tests/num`

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -930,6 +930,62 @@ $EndFeature, "
...
+assert_eq!(100", stringify!($SelfT), ".saturating_neg(), -100);
+assert_eq!((-100", stringify!($SelfT), ").saturating_neg(), 100);
+assert_eq!(", stringify!($SelfT), "::min_value().saturating_neg(), ", stringify!($SelfT),
+"::max_value());
+assert_eq!(", stringify!($SelfT), "::max_value().saturating_neg(), ", stringify!($SelfT),
+"::min_value() + 1);",
+$EndFeature, "
+```"),
+
+            #[unstable(feature = "saturating_neg", issue = "0")]
```
Comment (hellow554) [inline src/libcore/num/mod.rs]: Do we want to make #59983 the tracking issue or is a seperate issue desired?

Comment (t-rapp) [reply src/libcore/num/mod.rs]: My idea was to put the issue ID of this PR therein. If using #59983 is preferred I will happily use that.

Comment (hellow554) [reply src/libcore/num/mod.rs]: This was more a question to the reviewer. I'm not a core member and therefore cannot decide whether to use `0` for the moment or the Issue. But a *trackig issue* is always an **issue** number, never a PR.

**Comment by alexcrichton** [comment]
Thanks for the PR! I agree with @hellow554 that with some tests this is likely good to go!

For tracking issues we can either reuse the existing issue (and keep it open after this PR merges) or you can open a dedicated new issue for tracking these APIs, up to you!

**Comment by t-rapp** [comment]
Have added tests. I think we can use issue #59983 for tracking. Do I need to remove the "Fixes" line in the commit message so the tracking issue does not close? And what about merging trivial commit a616f10 into fc6272a ?

**Comment by hellow554** [comment]
> Do I need to remove the "Fixes" line in the commit message so the tracking issue does not close

either that or you manually reopen the issue afterwards ^^ (please somebody correct me if I'm wrong)
You can rebase your commits as you like, only merge commits are not wanted, the rest is okay-ish. As said, feel free to rebase/force push, if you want.

**Comment by t-rapp** [comment]
It seems I somehow managed to screw up the force-push as now commit 
0357e87 appears in the pull request. Sorry, I'm a newbie to GitHub. How can I get rid of that commit on the remote branch?

**Comment by hellow554** [comment]
Go back to the commit you started by using `git reset`, commit the changes (you have saved before of course) and do a `git push --force`. Let's hope that you don't forget anything :)

**Comment by t-rapp** [comment]
Thanks a lot for the hints! The PR looks good now, I think.

**Comment by alexcrichton** [comment]
Everything here looks good to me now, thanks again @t-rapp!

@bors: r+

**Pull request merged**